### PR TITLE
Add code to dump the image from virtual swapchain.

### PR DESCRIPTION
**Pull request by RenfengLiu**
Use environment variable to control  whether to dump the images from virtual swapchain to file. This is used for test the output of frameloop.

### CODE REVIEW

File: core/vulkan/vk_virtual_swapchain/cc/virtual_swapchain.cpp [hunk 1]
```diff
@@ -267,6 +279,87 @@ VirtualSwapchain::VirtualSwapchain(
...
+              << std::endl;
+    static int frame_count = 1;
+    ImageDumpInfo *dump_info = new ImageDumpInfo{
+        std::string(image_dump_path),
+        width_,
+        height_,
+        _swapchain_info->imageFormat,
+    };
+
+    SetCallback(
```
Comment (AWoloszyn) [inline core/vulkan/vk_virtual_swapchain/cc/virtual_swapchain.cpp]: What happens if the user wants to set a callback, but we have this env var set. that means we will end up overwriting it.

File: core/vulkan/vk_virtual_swapchain/cc/virtual_swapchain.cpp [hunk 2]
```diff
@@ -292,6 +385,11 @@ void VirtualSwapchain::Destroy(const VkAllocationCallbacks *pAllocator) {
   }
 
   functions_->vkDestroyCommandPool(device_, command_pool_, pAllocator);
+
+  if (callback_user_data_ != nullptr) {
```
Comment (AWoloszyn) [inline core/vulkan/vk_virtual_swapchain/cc/virtual_swapchain.cpp]: Same as above, if this comes from the user, we will be deleting a user object. That is very bad.

File: gapis/api/vulkan/frame_loop.go [hunk 3]
```diff
@@ -478,9 +478,13 @@ func (f *frameLoop) Transform(ctx context.Context, cmdId api.CmdID, cmd api.Cmd,
 
 				// Add unconditional jump instruction to bring us back to the start of the loop if we made it past the conditional break.
 				stateBuilder.write(stateBuilder.cb.Custom(func(ctx context.Context, s *api.GlobalState, b *builder.Builder) error {
-					b.Push(value.S32(1))
```
Comment (purvisa-at-google-com) [inline gapis/api/vulkan/frame_loop.go]: You plan on reverting the loop structure to before I fixed the last iteration rewind problem?

**Comment by RenfengLiu** [comment]
sorry, I pushed my testing code to the run branch. I comments that out as it seems has problem for destroy the resources.

**Comment by purvisa-at-google-com** [comment]
This might be related to a crash I have been seeing today. In theory, the old loop structure should be the one causing crashes, not the new one. Can you tell me about what you're seeing? Do you have a stack trace or an idea of which vkFunction that we're crashing in? What is the test app or trace?

**Comment by RenfengLiu** [comment]
The sample I usde is from SaschaWillems/Vulkan, with slight modification at https://github.com/RenfengLiu/Vulkan/commit/e82531bc0bb6ebb20fc8991146e13cd8ef667155#diff-3268f9dd8f19a1993cdccc85d1775373R606, which will casue the gapir crash after looping. there's much log in gapir for the crash, so that's why i'm adding logs in the tmpl files .

**Comment by purvisa-at-google-com** [comment]
I am working on a PR to add frame looping to replay export. You can then launch gapir directly on your host with a file representing the protos that GAPIS would have sent. If you launch gapir with verbose logging and GDB directly, you can learn a lot about where it broke. Will send you over repro steps soon!

**Comment by RenfengLiu** [comment]
yes, that would be very helpful for debugging.

**Pull request closed without merging**